addons/stock/tests/test_robustness.py

```diff
@@ -113,6 +114,53 @@ def test_location_usage(self):
         # unreserve
         move1._do_unreserve()
 
+    def test_internal_scrap_location_usage(self):
+        """Reservation should not be bypassed in internal scrap location."""
+
+        # change stock usage
+        test_stock_location = self.env['stock.location'].create({
+            'name': "Test Location",
+            'location_id': self.stock_location.id,
+            'usage': 'internal',
+        })
+        test_stock_location.scrap_location = True
+
+        # make some stock
+        self.env['stock.quant']._update_available_quantity(
+            self.product1,
+            test_stock_location,
+            1,
+        )
+
+        # reserve a unit
+        move1 = self.env['stock.move'].create({
+            'name': 'test_location_archive',
+            'location_id': test_stock_location.id,
+            'location_dest_id': self.customer_location.id,
+            'product_id': self.product1.id,
+            'product_uom': self.uom_unit.id,
+            'product_uom_qty': 1,
+        })
+        move1._action_confirm()
+        move1._action_assign()
+        self.assertEqual(move1.state, 'assigned')
+        quant = self.env['stock.quant']._gather(
+            self.product1,
+            test_stock_location,
+        )
+
+        # assert the reservation
+        self.assertEqual(quant.reserved_quantity, 1)  # reservation is not bypassed in scrap location
+        self.assertEqual(move1.product_qty, 1)
+
+        # change the stock usage
+        with self.assertRaises(UserError):
+            with self.cr.savepoint():
+                test_stock_location.scrap_location = False
+
+        # unreserve
```

Comment:
I am not sure if this part is required :thinking: Changes made in the database are reverted after each test, so unless there is a good reason to keep this line, I would remove it.

Reply:
done